### PR TITLE
implement update root

### PR DESCRIPTION
**Pull request by july2993**
implement update root

Seems we should VerifySignature according to the threshold of the manifest type but not just all signatures?
Also we should update the keys of `LocalRoot` when we get or save a new root manifest

### CODE REVIEW

File: pkg/repository/v1_repository.go [hunk 1]
```diff
@@ -88,13 +92,100 @@ func (r *V1Repository) updateLocalSnapshot() (*v1manifest.Snapshot, error) {
...
+	}
+
+	var newRoots []v1manifest.Manifest
+
+	for version := root1.Version + 1; ; version++ {
+		var root2 v1manifest.Root
+		fname := fnameWithVersion(v1manifest.ManifestURLRoot, version)
+
+		reader, err := r.mirror.Fetch(fname, int64(maxRootSize))
+		if err != nil {
```
Comment (lonng) [inline pkg/repository/v1_repository.go:117]: I think we break just if the error is `NotFound`, and other errors should propagate to the caller.

File: pkg/repository/v1_repository.go [hunk 2]
```diff
@@ -88,13 +92,100 @@ func (r *V1Repository) updateLocalSnapshot() (*v1manifest.Snapshot, error) {
...
+		return err
+	}
+
+	var newRoots []v1manifest.Manifest
+
+	for version := root1.Version + 1; ; version++ {
+		var root2 v1manifest.Root
+		fname := fnameWithVersion(v1manifest.ManifestURLRoot, version)
+
+		reader, err := r.mirror.Fetch(fname, int64(maxRootSize))
```
Comment (nrc) [inline pkg/repository/v1_repository.go:116]: Could `maxRootSize` be an `int64` and skip this conversion

File: pkg/repository/v1_repository.go [hunk 3]
```diff
@@ -88,13 +92,100 @@ func (r *V1Repository) updateLocalSnapshot() (*v1manifest.Snapshot, error) {
...
+			break
+		}
+
+		decoder := json.NewDecoder(reader)
+		var m v1manifest.Manifest
+		m.Signed = &root2
+		err = decoder.Decode(&m)
+		if err != nil {
+			return err
+		}
```
Comment (nrc) [inline pkg/repository/v1_repository.go:129]: I think decoding and basic verification should be done in manifest.go.

File: pkg/repository/v1manifest/manifest.go [hunk 4]
```diff
@@ -95,31 +95,50 @@ var ManifestsConfig = map[string]ty{
...
-// verifySignature ensures that each signature in manifest::signatures is a valid signature of manifest::signed.
-func (manifest *Manifest) verifySignature(keys crypto.KeyStore) error {
+// VerifySignature ensures that threshold signature in manifest::signatures is a valid signature of manifest::signed.
+func (manifest *Manifest) VerifySignature(threshold uint, keys crypto.KeyStore) error {
 	if keys == nil {
 		return nil
 	}
 
+	if threshold == 0 {
+		return nil
```
Comment (nrc) [inline pkg/repository/v1manifest/manifest.go:105]: This should be an error

File: pkg/repository/v1manifest/manifest.go [hunk 5]
```diff
@@ -95,31 +95,50 @@ var ManifestsConfig = map[string]ty{
...
+		// TODO use SignatureError
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			validCount++
 		}
 	}
 
-	return nil
+	if validCount >= threshold {
```
Comment (nrc) [inline pkg/repository/v1manifest/manifest.go:137]: I think all signatures must be correct, not just the threshold number. Threshold just dictates the number of signatures required.

**Pull request merged**